packages/langium/src/test/langium-test.ts

```diff
@@ -188,3 +189,120 @@ function replaceIndices(base: ExpectedBase): { output: string, indices: number[]
 
     return { output: input, indices, ranges: ranges.sort((a, b) => a[0] - b[0]) };
 }
+
+export interface ValidationResult<T extends AstNode = AstNode> {
+    diagnostics: Diagnostic[];
+    document: LangiumDocument<T>;
+}
+
+export function validationHelper<T extends AstNode = AstNode>(services: LangiumServices): (input: string) => Promise<ValidationResult<T>> {
+    const parse = parseHelper<T>(services);
+    return async (input) => {
+        const document = await parse(input);
+        return { document, diagnostics: await services.validation.DocumentValidator.validateDocument(document) };
+    };
+}
+
+export type ExpectDiagnosticOptionsWithoutContent<T extends AstNode = AstNode> = ExpectDiagnosticCode & (ExpectDiagnosticAstOptions<T> | ExpectDiagnosticRangeOptions | ExpectDiagnosticOffsetOptions);
+export type ExpectDiagnosticOptions<T extends AstNode = AstNode> = ExpectDiagnosticContent & ExpectDiagnosticOptionsWithoutContent<T>;
+
+export interface ExpectDiagnosticContent {
+    message?: string | RegExp
+    severity?: DiagnosticSeverity;
+}
+
+export interface ExpectDiagnosticCode {
+    code?: string;
+}
+
+export interface ExpectDiagnosticAstOptions<T extends AstNode> {
+    node: T;
+    property?: { name: Properties<T>, index?: number };
+}
+
+export interface ExpectDiagnosticRangeOptions {
+    range: Range;
+}
+
+export interface ExpectDiagnosticOffsetOptions {
+    offset: number
+    length: number
+}
+
+export type Predicate<T> = (arg: T) => boolean;
+
+function isRangeEqual(lhs: Range, rhs: Range): boolean {
+    return lhs.start.character === rhs.start.character
+        && lhs.start.line === rhs.start.line
+        && lhs.end.character === rhs.end.character
+        && lhs.end.line === rhs.end.line;
+}
+
+function filterByOptions<T extends AstNode = AstNode, N extends AstNode = AstNode>(validationResult: ValidationResult<T>, options: ExpectDiagnosticOptions<N>) {
+    const filters: Array<Predicate<Diagnostic>> = [];
+    if ('node' in options) {
+        const node = 'property' in options
+            ? findNodeForFeature(options.node.$cstNode, options.property!.name, options.property!.index)
+            : options.node.$cstNode;
```

Comment:
This can actually fail, hence why TypeScript complains about `property` being possibly undefined: `'x' in { x: undefined }` returns true, since a property `x` exists, even though it is undefined.

Also, naming this also `node` is quite ambiguous, since we have one ast node and a cst node.

```suggestion
        const cstNode = options.property
            ? findNodeForFeature(options.node.$cstNode, options.property.name, options.property.index)
            : options.node.$cstNode;
```